src/words/fast.jl

```diff
@@ -0,0 +1,193 @@
+
+TokenBuffer(input::AbstractString) = TokenBuffer(collect(input))
+
+Base.getindex(ts::TokenBuffer, i = ts.idx) = ts.input[i]
+isdone(ts::TokenBuffer) = ts.idx > length(ts.input)
+
+"""
+    flush!(::TokenBuffer, tokens...)
+
+TokenBuffer builds the current token as characters are read from the input. When
+the end of the current token is detected, call `flush!` to finish it and append
+it to the token stream. Optionally, give additional tokens to be added to the
+stream after the current one.
+"""
+function flush!(ts::TokenBuffer, s...)
+  if !isempty(ts.buffer)
+    push!(ts.tokens, String(ts.buffer))
+    empty!(ts.buffer)
+  end
+  push!(ts.tokens, s...)
+  return
+end
+
+"""
+    lookahead(::TokenBuffer, s; boundary = false)
+
+Peek at the input to see if `s` is coming up next. `boundary` specifies whether
+a word boundary should follow `s`.
+
+    julia> lookahead(TokenBuffer("foo bar"), "foo")
+    true
+
+    julia> lookahead(TokenBuffer("foo bar"), "bar")
+    false
+
+    julia> lookahead(TokenBuffer("foo bar"), "foo", boundary = true)
+    true
+
+    julia> lookahead(TokenBuffer("foobar"), "foo", boundary = true)
+    false
+"""
+function lookahead(ts::TokenBuffer, s; boundary = false)
+  ts.idx + length(s) - 1 > length(ts.input) && return false
+  for j = 1:length(s)
+    ts.input[ts.idx-1+j] == s[j] || return false
+  end
+  if boundary
+    next = ts.idx + length(s)
+    next > length(ts.input) && return true
+    (isletter(ts[next]) || ts[next] == '-') && return false
+  end
+  return true
+end
+
+"""
+    character(::TokenBuffer)
+
+Push the next character in the input into the buffer's current token.
+"""
+function character(ts)
+  push!(ts.buffer, ts[])
+  ts.idx += 1
+  return true
+end
+
+"""
+    spaces(::TokenBuffer)
+
+If there is whitespace in the input, skip it, and flush the current token.
+"""
+function spaces(ts)
+  isspace(ts[]) || return false
+  flush!(ts)
+  ts.idx += 1
+  return true
+end
+
+"""
+    atoms(::TokenBuffer, ["--", "...", ...])
+
+Matches a set of atomic tokens, such as `...`, which should always be treated
+as a single token, regardless of word boundaries.
+"""
+function atoms(ts, as)
+  for a in as
+    lookahead(ts, a) || continue
+    flush!(ts, String(a))
+    ts.idx += length(a)
+    return true
+  end
+  (ispunct(ts[]) && ts[] != '.' && ts[] != '-') || return false
+  flush!(ts, string(ts[]))
+  ts.idx += 1
+  return true
+end
+
+"""
+    suffixes(::TokenBuffer, ["'ll", "'re", ...])
+
+Matches tokens with suffixes, such as `you're`, that should be treated as
+separate tokens.
+"""
+function suffixes(ts, ss)
+  isempty(ts.buffer) && return false
+  for s in ss
+    lookahead(ts, s, boundary=true) || continue
+    flush!(ts, String(s))
+    ts.idx += length(s)
+    return true
+  end
+  return false
+end
+
+"""
+    suffixes(::TokenBuffer, [("cannot", 3), ("gimme", 3), ...])
+
+Matches tokens that should be split at the given index. For example, `cannot`
+would be split into `can` and `not`.
```

Comment:
```suggestion
would be split into `can` and `not`. i.e. at the 3rd index.
So we pass in `("cannot", 3)`
```

As an aside.
I think it might be both faster, and cleaner
to have `replacers(ts, ("cannot" => ("can", "not")), ("gimme" => ("gim", "me")))`
instead.

What do you think?